### PR TITLE
add results to community hub contract

### PR DESCRIPTION
**Pull request by jordipainan**
Some questions I have:

- Is it fine to set the date equal to the date where the election finished ?
- Do we really want to store the turnout as it is? I need to truncate it cause solidity does not support floats
- How I can get the total voting power from the election || electiondb struct ?
- Do we really want to store the list of participants? How I can get it from the election || electiondb struct ?

### CODE REVIEW

File: results.go [hunk 1]
```diff
@@ -122,6 +124,34 @@ func (v *vocdoniHandler) finalizeElectionResults(election *api.Election, electio
 			return
 		}
 		log.Infow("final results image built ondemand", "electionID", election.ElectionID.String())
+		hubResults := &communityhub.HubResults{
```
Comment (p4u) [inline results.go:127]: You should check if the election is part of a community that exist on the smart contract.

File: results.go [hunk 2]
```diff
@@ -122,6 +124,34 @@ func (v *vocdoniHandler) finalizeElectionResults(election *api.Election, electio
...
+			CensusURI:  election.Census.CensusURL,
+		}
+		hubResults.Tally = make([][]*big.Int, 0)
+		for _, option := range election.Results[0] {
+			opt := new(big.Int).Set(option.MathBigInt())
+			hubResults.Tally[0] = append(hubResults.Tally[0], opt)
+		}
+		// hubResults.Turnout = big.NewInt(int64(electiondb.Turnout))
+		// hubResults.TotalVotingPower = big.NewInt(int64(election.Census.MaxCensusSize))
+		// hubResults.Participants =
```
Comment (p4u) [inline results.go:141]: Why not add participants? What's the maximum size of participants we can add? Maybe we can check for a maximum and do it only if it's not bigger.

File: results.go [hunk 3]
```diff
@@ -122,6 +124,34 @@ func (v *vocdoniHandler) finalizeElectionResults(election *api.Election, electio
...
+			CensusRoot: election.Census.CensusRoot,
+			CensusURI:  election.Census.CensusURL,
+		}
+		hubResults.Tally = make([][]*big.Int, 0)
+		for _, option := range election.Results[0] {
+			opt := new(big.Int).Set(option.MathBigInt())
+			hubResults.Tally[0] = append(hubResults.Tally[0], opt)
+		}
+		// hubResults.Turnout = big.NewInt(int64(electiondb.Turnout))
+		// hubResults.TotalVotingPower = big.NewInt(int64(election.Census.MaxCensusSize))
```
Comment (p4u) [inline results.go:140]: I don't think we have the total voting power, so we should store it somewhere when creating the census.

File: results.go [hunk 4]
```diff
@@ -122,6 +124,34 @@ func (v *vocdoniHandler) finalizeElectionResults(election *api.Election, electio
...
+			Date:       election.EndDate.String(),
+			CensusRoot: election.Census.CensusRoot,
+			CensusURI:  election.Census.CensusURL,
+		}
+		hubResults.Tally = make([][]*big.Int, 0)
+		for _, option := range election.Results[0] {
+			opt := new(big.Int).Set(option.MathBigInt())
+			hubResults.Tally[0] = append(hubResults.Tally[0], opt)
+		}
+		// hubResults.Turnout = big.NewInt(int64(electiondb.Turnout))
```
Comment (p4u) [inline results.go:139]: Actually, turnout should be computed from total voting power, so the one in the database is not correct. We should fix that.

File: results.go [hunk 5]
```diff
@@ -122,6 +124,34 @@ func (v *vocdoniHandler) finalizeElectionResults(election *api.Election, electio
...
+		// hubResults.TotalVotingPower = big.NewInt(int64(election.Census.MaxCensusSize))
+		// hubResults.Participants =
+		if err := v.comhub.SetResults(electiondb.Community.ID, election.ElectionID, hubResults); err != nil {
+			log.Errorw(err, "failed to set results on the community hub")
```
Comment (p4u) [inline results.go:143]: If it fails, shouldn't we exit?

File: results.go [hunk 6]
```diff
@@ -122,6 +124,34 @@ func (v *vocdoniHandler) finalizeElectionResults(election *api.Election, electio
...
+		}
+		// check results added to the community hub
+		for i := 0; i < 3; i++ {
```
Comment (p4u) [inline results.go:146]: What's the reason for this check? I think SetResults() if no error returned, should either be sure it worked fine or run this check internally .

**Review by p4u — CHANGES_REQUESTED** [review]

File: results.go (hunk 2 above)
Comment (jordipainan) [reply results.go:141]: there is no limit in size, but the smart contract needs to be changed as it is expecting fids and not usernames

File: results.go (hunk 5 above)
Comment (jordipainan) [reply results.go:143]: yup

File: results.go (hunk 6 above)
Comment (jordipainan) [reply results.go:146]: It was just to check the tx was mined. But can be avoided.

**Pull request merged**